design/one-pager-fn-claim-conditions.md

```diff
@@ -0,0 +1,170 @@
+
+### Scenarios
+Given a few different scenarios, users could expect to see the following `status.conditions` for
+the claim.
+
+#### Image Not Found
+First we found the database and determined that the user has authorization, however, the image they
+provided was not found.
+
+An example of the Claim's status:
+```yaml
+status:
+  conditions:
+  - type: DatabaseReady
+    status: True
+    reason: Available
+  - type: ImageReady
+    status: False
+    reason: NotFound
+    message: The image provided does not exist or you are not authorized to use it.
+  - type: AppReady
+    status: Unknown
+    reason: PreviousErrors
+    message: There were previous errors which prevented us from updating this condition.
+```
+#### Progressing
+All is fine and the application is progressing but not yet fully online.
+
+An example of the Claim's status:
+```yaml
+status:
+  conditions:
+  - type: DatabaseReady
+    status: True
+    reason: Available
+  - type: ImageReady
+    status: True
+    reason: Available
+  - type: AppReady
+    status: False
+    reason: Creating
+    message: Waiting for the deployment to be available.
+```
+
+#### Success
+Once everything is online and running smoothly, users should see something like this.
+
+An example of the Claim's status:
+```yaml
+status:
+  conditions:
+  - type: DatabaseReady
+    status: True
+    reason: Available
+  - type: ImageReady
+    status: True
+    reason: Available
+  - type: AppReady
+    status: True
+    reason: Available
+```
+
+### Team Implementation
+To accomplish this behavior, the team would need to configure their function pipeline to have the following
+behavior.
+1. The first step of the pipeline must always "reserve" the expected `status.conditions` on the desired XR.
+  In this example above, this would be to create an entry for all three condition types
+  (`DatabaseReady`, `ImageReady`, `AppReady`) and set each to a default of:
+  ```yaml
+    - type: <type>
+      status: Unknown
+      reason: PreviousErrors
+      message: There were previous errors which prevented us from updating this condition.
+  ```
+  This is required in the case that we exit early due to an error. If we did not pre-populate this and we hit an error
+  before creating an entry, the Claim will remove that condition from it's status, assuming the condition is no longer
+  desired.
+1. As we reach points in the function where we wish to update a specific condition, we can do so.
+  ```go
+  c := xpv1.Condition{...}
+  desiredXR.Resource.SetConditions(c)
+  response.SetDesiredCompositeResource(rsp, desiredXR)
+  ```
+
+## Alternatives Considered
+
+### Events
+In our search for providing communication to the Claim, we considered giving Composition Function 
+authors the ability to send events to the Claim, however, we believe the proposal above is preferred
+as it provides the ability to communicate with users in a more structured way.
```

Comment:
I might have been unclear WRT my position on events yesterday.

I don't think we need to _load_ events into the function pipeline, i.e. send composed resource events to functions in `RunFunctionRequest`. I could imagine doing that in future to give function authors more context they could use to set status conditions and/or produce results, but I would defer that until we know we need it.

I do like the idea of allowing function authors to emit "claim aligned" results (i.e. events). So something like:

```protobuf
// A Result of running a Function.
message Result {
  // Severity of this result.
  Severity severity = 1;

  // Human-readable details about the result.
  string message = 2;

  // Targets of the result. The result will be associated with each target.
  // If no targets are specified, the result targets  only the composite.
  repeated Target targets = 3;
}

// Target of Function results.
enum Target {
  TARGET_UNSPECIFIED = 0;
  TARGET_COMPOSITE = 1;
  TARGET_CLAIM = 2;
}
```

What I'm not sure about is how results should interact with status conditions.

Reply:
Maybe there's a hybrid approach between what I sketched out in the preceding comment and what's currently in the proposal:

* Let results explicitly target an XR, claim, or both.
* Let results optionally update a status condition, in addition to emitting an event.
* If the pipeline produces >1 result targeting the same status condition, only the last one is set (the others are just events).

So something like this:

```protobuf
// A Result of running a Function.
message Result {
  // Severity of this result.
  Severity severity = 1;

  // Human-readable details about the result.
  string message = 2;

  // Machine-readable PascalCase reason for the result.
  string reason = 3;

  // Targets of the result. The result will be associated with each target.
  // If no targets are specified, the result targets only the composite.
  repeated Target targets = 4;

  // Optionally update the supplied status condition on all targets.
  // The result's reason and message will be used in the condition.
  optional Condition condition = 5;
}

// Target of Function results.
enum Target {
  TARGET_UNSPECIFIED = 0;
  TARGET_COMPOSITE = 1;
  TARGET_CLAIM = 2;
}

message Condition {
  // Type of the condition, e.g. DatabaseReady.
  // 'Ready' and 'Synced' are reserved for use by Crossplane.
  string type = 1;

  // Status of the condition.
  Status status = 2;
}

// Status of Function result condition.
enum Target {
  STATUS_UNSPECIFIED = 0;
  STATUS_FALSE = 1;
  STATUS_TRUE = 2;
}
```

Reply:
Note that `Target` would specifically refer to the receiver of the `Event`. The status condition would need to be set on the XR if we want the Claim to receive them. Unless if we want to create another way to copy the conditions from the XR to the Claim, such as creating a field on the XR specifically for claim-only conditions.

Reply:
> Let results optionally update a status condition, in addition to emitting an event.

IMO we should also enable the user to set a condition without explicitly setting an event. The user could let Crossplane decide when to emit events.

We could have logic at the reconciler that emits an event if:
`condition_is_new || condition_has_changed`

`condition_is_new` means the condition.Type was not seen as part of the previous conditions.
`condition_has_changed` means the condition.Type was seen previously, but the condition.Status has changed.

Reply:
> Note that Target would specifically refer to the receiver of the Event. The status condition would need to be set on the XR if we want the Claim to receive them.

I'm proposing that the target should refer to the receiver of the event, _and_ of the status condition (if there is one).

So for example if a function emits a result with `TARGET_CLAIM` and a `Condition` message, I would propose that event and status condition _only_ appear on the claim, not on the XR.

> Unless if we want to create another way to copy the conditions from the XR to the Claim, such as creating a field on the XR specifically for claim-only conditions.

Right. Admittedly my proposal is focused on my preferred UX, and ignores how we'd actually build it. Of all the ways we could have a function pipeline influence claim status conditions, what you suggest seems like it would probably be the least bad.

The other alternatives that came to mind, that I like less:

* Have the XR controller mutate claims directly.
* Have the XR controller emit events about claims, and infer status conditions from those.

Reply:
> how we'd actually build it

Maybe we eliminate `TARGET_CLAIM` only as an option? So more like:

```proto
message Result {
  // Omitted for brevity

 Target target = 4;
}

// Target of Function results.
enum Target {
  TARGET_UNSPECIFIED = 0;
  TARGET_COMPOSITE_ONLY = 1;
  TARGET_COMPOSITE_AND_CLAIM = 2;
}
```

We could then keep a mask of XR status conditions that should propagate to the claim.

My thinking is that if the point of claim-aligned status conditions and events is to produce user-friendly information, it doesn't hurt anything for that user-friendly information to appear on the XR too.

Reply:
> IMO we should also enable the user to set a condition without explicitly setting an event. The user could let Crossplane decide when to emit events.

That would be a departure from what we do today. I believe:

* Every status condition transition produces an event (a record of the status change).
* Some things produce only events.

That's what I was going for with my proposed API:

* Every result is an event.
* Some results also change the status condition.

Reply:
>That would be a departure from what we do today.

I'm just saying to provide a mechanism to turn off "Every result is an event".
If this option is turned off by the user, then "Every status condition transition produces an event (a record of the status change)." would take care of emitting the event.

By default, every result would be an event (e.g., calling `response.Fatal`). But if a user opts into a new API (e.g., `response.Condition`), then by default we would not emit the event unless the condition has changed (possibly no new code, just letting the existing condition logic take effect).

Reply:
I updated the doc to contain some of your code snippets and examples of the function author creating events and conditions that target the claim.

Reply:
> I am just thinking that in the case of results containing conditions, it may be nice to reduce event count if conditions are not changing.

Yeah good call - we don't want to emit an event if it's effectively just capturing that the status is unchanged.